service/src/main/java/org/broadinstitute/listener/relay/http/RelayedHttpRequestProcessor.java

```diff
@@ -83,6 +97,38 @@ public Result writeNotAcceptedResponseOnCaller(RelayedHttpListenerContext contex
     return Result.FAILURE;
   }
 
+  public Result writePreflightResponse(RelayedHttpListenerContext context) {
+    if (context.getResponse() == null) {
+      logger.error("The context did not have a valid response");
+      return Result.FAILURE;
+    }
+
+    RelayedHttpListenerResponse listenerResponse = context.getResponse();
```

Comment:
maybe is another ticket, but we need to validate that the CORS request is valid - contains the expected headers and the requested URL is checked against a allowed list/setting - before creating a valid response.

Reply:
I think that's basically the same as `Referer` header check... @dinvlad is it necessary to check both `Referer` header and `Origin` header?

Reply:
I would test it - I believe all browsers but IE provide `Referer`, but we may want to check `Origin` also if `Referer` is not available (but `Origin` may not be provided, based on the context, so please test it in the field)

Reply:
A few more details/thoughts on this: `Origin` is a required header as per the CORS algorithm spec details [here](https://www.w3.org/TR/2020/SPSD-cors-20200602/#resource-preflight-requests).  So if it is missing, or not valid, the request must be terminated.  A `Referer` check is outside the scope of CORS from what I can tell.  If we have a functional need to check the `Referer`, IMO it is best to implement the check as an `inspector` and not part of the CORS validation process.

Reply:
I think whichever option works across browsers, it doesn't matter that much. Please use this as a reference wrt CSRF protection: https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#verifying-origin-with-standard-headers

Reply:
Also, if Origin works then it might be slightly _preferred_, because without a proper Referrer-Policy header, Referer might contain URL path in addition to the origin itself, and then the path needs to be parsed out, which adds room for error. Better keep it simple with Origin, if possible.